오픈마켓 [STEP 2] SummerCat

ttokin opened the pull request:
안녕하세요 엘림! @lina0322 

`listCollectionView`는 Compositional Layout으로, `gridCollectionView`는 Flow Layout으로 구현했고, 데이터 소스는 Diffable Data Source를 활용했습니다. 프로젝트의 일관성을 위해서는 한 가지 레이아웃으로 통일하는게 좋겠지만, 컬렉션 뷰 구현이 처음이라 둘 다 구현해보는 데 더 초점을 두고 진행했습니다.

## 고민했던 점

#### 1. 가격 Label의 개수와 무관하게 상품명 Label이 일정한 위치를 유지하도록 하기
`stackView`에 `productImageView, productNameLabel, priceStackView(priceLabel, bargainPriceLabel), stockLabel`이 담겨 있을 때에는 `priceLabel`과 `bargainPriceLabel` 중 하나만 표시할 경우 스택 뷰의 설정이 `.equalSpacing`으로 되어있어 `productNameLabel`의 위치가 살짝 아래로 조정되었습니다. 이는 사용자 경험을 해칠 수 있다고 생각해 `productNameLabel`이 아래 사진과 같이 일정한 위치를 유지할 수 있는 방법을 고민했고, 조언을 받아 가로 스택 뷰 안에 `priceStackView`를 넣어 `priceStackView`에 고정된 높이를 줄 수 있도록 변경하였습니다. 세로 스택 뷰의 경우 세로 크기를 정해줄 수 없고 가로 크기를 정해줄 수 있는 반면, 가로 스택 뷰의 경우 세로 크기를 정해줄 수 있기 때문입니다.

![](https://i.imgur.com/wLodQgB.png)

#### 2. 포함하고 있는 텍스트 콘텐트가 매우 긴 Label을 어떻게 처리할지
아래 예시의 경우 상품명, 가격, 재고 Label 모두 긴 텍스트 콘텐트를 포함하고 있습니다.
재고 숫자가 커서 Label이 길어질 경우 "재고 여유 있음"이라는 텍스트로 수량을 대체했습니다.

![](https://i.imgur.com/eMHh25G.png)

그리드 뷰의 경우 텍스트가 길어지면 2줄로 나누어서 표시해줄지, 위 이미지와 같이 truncate tail 형태로 표시할지 고민하다가 위 이미지와 같이 표시하기로 했습니다. 두 줄로 표시할 경우 텍스트 사이즈가 작아지고, 한 줄짜리 제목과 두 줄짜리 제목이 나란히 있을 경우 UI의 일관성이 떨어진다고 생각해서입니다.

![](https://i.imgur.com/mwoFiIw.png)

리스트 뷰의 경우 셀의 크기가 작기 때문에 텍스트를 두 줄로 표시하는 것은 배제하고 구현하였습니다.


## 조언을 얻고 싶은 부분

#### 1. 빠르게 스크롤해서 셀이 재사용될 때 셀 내의 콘텐츠가 잠시 다른 콘텐츠가 표시되었다가 원래 표시되어야 할 콘텐츠가 표시되는 문제
![](https://i.imgur.com/iXSf7sn.gif)

위 화면과 같이 셀이 재사용되면서 이전에 가지고 있던 이미지, Label 등을 보여주다가 해당 셀에 적합한 콘텐츠를 표시하는 문제를 해결하기 위해, 각 셀에 `var product: Product`라는 프로퍼티를 추가했습니다. 그리고 셀을 dequeue할 때 셀의 `product` 프로퍼티의 값을 할당하고, 셀의 콘텐츠를 업데이트 하는 코드 내에서 인자로 넘겨받은 `product`와 프로퍼티인 `product`를 비교해 같을 때에만 셀의 콘텐츠를 업데이트 하는 것으로 조건을 주었습니다.

```swift
func updateContents(_ product: Product) {
    DispatchQueue.global().async {
        guard let url = URL(string: product.thumbnailURL),
              let data = try? Data(contentsOf: url),
              let image = UIImage(data: data) else { return }
        DispatchQueue.main.async {
            if product == self.product {
                self.productImageView.image = image
                self.productNameLabel.text = product.name
                self.updatePriceLabel(product)
                self.updateStockLabel(product)
            }
        }
    }
}
```

실행해보면 로딩이 조금 느리긴 하지만 정상적으로 작동하는 것 같은데.. 제대로 된 검증 방법일지, 더 나은 검증 방법이 있을지 궁금합니다!

#### 2. 역할이 비슷한 코드 합치기
아래의 코드는 `segmentControl`의 값에 따라 리스트 컬렉션 뷰 / 그리드 컬렉션 뷰 중 하나를 선택해서 선택한 뷰를 설정하는 메서드와 각 뷰에 맞는 데이터 소스를 설정하는 메서드를 호출하고 있습니다. 그런데 다소 중복되는 부분도 있고.. 더 좋은 방향으로 리팩토링 할 여지가 있다고 생각되는데 어떻게 하면 좋을지 방법이 잘 떠오르지 않아 엘림의 의견이 궁금합니다!
그리고 현재 코드에서 이런(역할이 비슷한 코드가 여러 개 있는) 부분이 혹시 더 있을까요??

```swift
@objc private func segmentValueChanged(_ sender: UISegmentedControl) {
    guard let viewType: ViewType = ViewType(rawValue: sender.selectedSegmentIndex) else { return }

    switch viewType {
    case ViewType.list:
        configureListCollectionView()
        configureListDataSource()
        applySnapshot(for: products)
        listCollectionView?.isHidden = false
        gridCollectionView?.isHidden = true
    case ViewType.grid:
        configureGridCollectionView()
        configureGridDataSource()
        applySnapshot(for: products)
        gridCollectionView?.isHidden = false
        listCollectionView?.isHidden = true
    }
}
```


#### 3. 스크롤 시 로그에 찍히는 에러 로그
리스트 뷰와 그리드 뷰 모두 스크롤 하다 보면 아래와 같은 에러 로그가 찍히는데, 구글링해보아도 해당 로그에 대한 정보가 거의 없어 혹시 엘림은 짐작가시는 바가 있는지 여쭈어봅니다..!

![](https://i.imgur.com/RlUj50z.png)
```text
2022-12-01 01:46:44.553300+0900 OpenMarket[88898:5754063] [connection] nw_connection_add_timestamp_locked_on_nw_queue [C3] Hit maximum timestamp count, will start dropping events

```

#### 4. listCollectionViewCell 오토레이아웃 이슈
스택 뷰 여러 개를 사용해 리스트 형태의 셀을 구현했는데, 시뮬레이터 실행 시 화면은 오류 없이 작동하지만 제약조건 충돌(정확히는 제약조건 부족)로 인해 오류가 계속 발생하고 있습니다... 혼자서 열심히 고쳐보려고 했지만 고치지 못해 도움 요청드립니다.ㅠㅠ

hyhy0429 committed:
feat: GridCollectionViewCell 타입 구현

ttokin committed:
feat: Formatter Extension 생성

hyhy0429 committed:
feat: GridCollectionViewCell 구현

ttokin committed:
feat: OpenMarketVC에 gridCollectionView 추가 및 DataSource 생성

ttokin committed:
refactor: GridCollectionView의 view를 lazy var에서 let으로 수정, label의 text를 전부 attributedText로 수정, format 메서드의 반환값 수정, productList URL queryParameter에서 itemsPerPage 제거

ttokin committed:
feat: willDisplay cell을 이용한 pagination 구현

ttokin committed:
feat: ListCollectionViewCell 구현, listCollectionView 구현

ttokin committed:
feat: segmentedControl 구현

ttokin committed:
refactor: listCollectionView cell 레이아웃 수정

ttokin committed:
refactor: 매직 리터럴 제거, stockLabel attributedText로 수정

ttokin committed:
feat: 상품 추가 화면 구현

ttokin committed:
feat: activityIndicator 구현

ttokin committed:
chore: 주석 제거 및 가독성 향상

ttokin committed:
feat: priceStackView를 감싸는 priceHStackView 구현

ttokin committed:
feat: 각 셀 타입에 product 프로퍼티 추가, cell의 product와 updateContents의 product를 비교하는 로직 추가

ttokin committed:
chore: fetchData 메서드 내 개행 추가, ListCollectionViewCell의 priceLabel, bargainPriceLabel 크기 조정

ttokin committed:
refactor: Formatter의 extension을 Double의 extension으로 변경, formatToDecimal 메서드 구현

ttokin committed:
refactor: listCollectionViewCell 리팩토링

ttokin committed:
refactor: Double의 formatting 메서드 반환타입 옵셔널로 변경, listCollectionViewCell 제약조건 수정, StockStatus에 case 추가

lina0322 commented on OpenMarket/OpenMarket/Extension/Double+.swift:
```diff
@@ -0,0 +1,32 @@
+
+import Foundation
+
+extension Double {
```
extension을 아주 잘 활용했네요!👍

lina0322 commented on OpenMarket/OpenMarket/Extension/String+.swift:
```diff
@@ -0,0 +1,19 @@
+        attributeString.addAttribute(NSAttributedString.Key.foregroundColor, value: UIColor.orange, range: NSMakeRange(0, attributeString.length))
+        return attributeString
+    }
+}
```
보기가 어려워요ㅠ 적당한 줄바꿈ㅇㅣ 필요해보입니다!

ttokin replied:
c53b9334a79e3de611b4191d9ce66883645a4695

제 나름대로 컨벤션에 맞추어 수정해보았습니다..!

lina0322 replied:
수정한 부분 좋습니다! 
다만 최소 버전이 변경되는 중대한(?)일이 하나의 커밋에 함께 되어있네요ㅎㅎㅎ
커밋 분리하는게 더 좋았을 것 같아요, 참고만 해주세요!

lina0322 commented on OpenMarket/OpenMarket/Model/Product.swift:
```diff
@@ -4,22 +4,22 @@
-struct Product: Decodable, Identifiable, Equatable {
-    let ID: Int
+struct Product: Decodable, Identifiable, Hashable {
+    let id: Int
```
Equatable을 Hashable 변경한 이유를 설명해주실 수 있나요~?🙏

ttokin replied:
57bcb0a4d69f50aea6ba32fe69779d78cf28536e

`Hashable`을 채택하도록 변경한 이유는 diffable data source의 `itemIdentifierType`을 `Product` 타입의 특정 인스턴스 프로퍼티(예: `id`)가 아닌 `Product`타입으로 지정했기 때문입니다. `Product`가 `Hashable`을 채택하면 hash function을 통해 hash value를 생성해서 hash value를 통해 `Product`의 각 인스턴스를 구분할 수 있게 됩니다. 
`Hashable`은 `Equatable`을 채택하도록 정의되어 있으며, `Product` 안의 모든 프로퍼티는 `Equatable`하기 때문에(구조체나 열거형은 프로퍼티가 모두 Equatable 할 경우 Equatable을 따로 채택하지 않아도 Equatable하기 때문) `Hashable`만 채택하고 따로 `==` 함수를 구현해주지 않아도 됩니다. 

```swift
private var dataSource: UICollectionViewDiffableDataSource<ProductListSection, Product>?
```
만약 `Product`의 인스턴스가 아닌 인스턴스 프로퍼티 `id`를 `itemIdentifierType`으로 지정하면 어떨지 궁금해서 위 코드에서 `Product` 뒤에 `.id`를 넣어보려고 했는데 `.ID`만 나오더라구요..!!
공식 문서의 Diffable Data Source를 적용하는 부분을 다시 공부해보고 엘림이 지난번에 말씀해주신 `Identifiable` 프로토콜을 채택하면 해당 프로토콜을 채택한 타입의 `id` 프로퍼티를 `.ID`로 가져와 쓰게 된다는 것을 알게 되었습니다..!

공식문서의 예제를 참고해서 Diffable Data Source와 Snapshot이 이제는 `Product`의 인스턴스 자체가 아닌 `Product.ID`를 저장하도록 변경해 보았습니다. identifier를 이용해서 원하는 상품을 찾으려면 `products`라는 배열을 돌면서 같은 `id` 값을 가진 요소를 탐색해야 하니 O(N)으로 탐색 시간이 늘어나지 않을까? 생각했었는데, 기존에도 객체끼리 비교해서 같은 객체를 탐색하기 때문에 똑같이 탐색시간이 O(N)일 것 같아요.

lina0322 replied:
효율성을 고려하여 코드를 작성한 시도는 좋았던 것 같아요! 
근데 dataSource 생성할때나 dataSource에서 item을 꺼내올때 Product로 값이 나오지 않으니, 
그때마다 id로 배열에서 검색하여 사용하는 번거로움이 생긴 것 같은데용.
이 부분은 어떻게 생각하실까요??

(혹시 어떤 공식문서였는지 알려주실 수 있나용?👀)

ttokin replied:
코드를 천천히 보면서 다시 생각해보니 `Product`의 인스턴스 자체를 가져오게 할 경우에는 Diffable Data Source가 identifier를 key로 이용해서 값을 가져오니 O(1)일 것 같기도 하네요🤔 이게 맞다면 `ID`로 가져오게 되면 배열을 탐색해야 해서 추가로 O(N)의 탐색시간이 필요하게 되겠네요.
대신 `ID`를 사용하면 데이터소스나 스냅샷을 더 가볍게 유지할 수 있지 않을까요..?

공식문서 예제는 아래 링크 참고했습니다!
https://developer.apple.com/documentation/uikit/views_and_controls/collection_views/updating_collection_views_using_diffable_data_sources

lina0322 replied:
섬머캣이 확인한 공식문서에 이런 내용이 있네욤?👀
<img width="697" alt="스크린샷 2022-12-06 11 48 12" src="https://user-images.githubusercontent.com/49546979/205797194-0abe3056-11d9-467f-8b0e-29252f554351.png">

ttokin replied:
맞아영..! 이 내용 읽고서 현재 코드와 같은 형태로 수정했습니다!
 구현 예제를 보니까 데이터소스/스냅샷에서는 `ID`를 사용해서 identifier를 가져온 다음에, 실제 데이터의 인스턴스를 담아둔 배열을 탐색해서 identifer와 같은 `ID` 값을 가진 인스턴스를 가져오는 메서드를 따로 구현해서 사용하고 있더라구요.
인스턴스가 같은지 비교하는 것보다 ID만 비교하는 게 더 빨라서 퍼포먼스가 낫다는 걸까요..?

lina0322 replied:
음, 그런 의미랑 비슷한 것 같아요. 굳이 hashable을 할 필요가 없이 내부에 ID 값은 hashableㅎㅏ니까 그 값을 사용하면 된다는 의미라고 저는 이해했어요ㅎㅎ

lina0322 commented on OpenMarket/OpenMarket/Info.plist:
```diff
@@ -43,8 +41,6 @@
 	<string>LaunchScreen</string>
-	<key>UIMainStoryboardFile</key>
-	<string>Main</string>
 	<key>UIRequiredDeviceCapabilities</key>
```
화면을 코드로 구성하신 이유가 궁금합니다!

ttokin replied:
지난 프로젝트인 Bank Manager 때 코드로 화면 구현해보기 부분을 해보지 못해서 이번에는 코드로 구현하는 것을 목표로 진행해 보았습니다.

lina0322 replied:
코드로 화면 짜기 vs 스토리보드로 화면 짜기

각각의 장단점을 한번 정리해보면 나중에 면접때 도움이될지두..🥹

lina0322 commented on OpenMarket/OpenMarket/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,245 @@
+    private var listCollectionView: UICollectionView?
+            
+    private let segmentedControl: UISegmentedControl = {
+        let control = UISegmentedControl(items: [ViewType.list.typeName, ViewType.grid.typeName])
```
item이 더 많아지면 하나씩 수동으로 넣어줘야겠네용?👀

ttokin replied:
segmentedControl을 설정하는 부분을 메서드화 하고, segment를 추가하는 함수를 추가로 구현해 보았습니다.

lina0322 commented on OpenMarket/OpenMarket/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,245 @@
+            
+    private let segmentedControl: UISegmentedControl = {
+        let control = UISegmentedControl(items: [ViewType.list.typeName, ViewType.grid.typeName])
+        control.translatesAutoresizingMaskIntoConstraints = false
```
translatesAutoresizingMaskIntoConstraints은 왜 false로 설정할까요??

ttokin replied:
오토레이아웃 제약을 직접 설정해주기 위해서 `false`로 설정하는데 (`true`로 설정하면 오토리사이징이 적용되서 오토레이아웃 제약과 충돌 발생), 위의 세그먼트 컨트롤 같은 경우는 설정된 오토레이아웃 제약이 없어서 `false`로 변경할 필요가 없네요..! 해당 부분은 삭제 처리했습니다!

lina0322 commented on OpenMarket/OpenMarket/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,245 @@
+    }
+    
+    private var gridCollectionView: UICollectionView?
+    private var listCollectionView: UICollectionView?
```
listCollectionView도 collectionView로 한 이유는 학습때문이죠?ㅎㅎ

ttokin replied:
넵 맞습니다..! 테이블뷰를 사용할 수도 있지만 기존 컬렉션뷰와 모던 컬렉션뷰 모두 학습해보고 싶어서요!

lina0322 commented on OpenMarket/OpenMarket/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,245 @@
+    private var dataSource: UICollectionViewDiffableDataSource<ProductListSection, Product>?
+    private var products: [Product] = [] {
+        didSet {
+            DispatchQueue.main.async { [weak self] in
+                self?.applySnapshot(for: self?.products)
+                self?.activityIndicator.stopAnimating()
+            }
```
1. 함수로 분리해주세요.
2. products가 변경됐을때 이렇게 변화를 주는 방식을 택한 이유가 궁금합니다. 변경하는 그 시점에 이 부분을 마찬가지로 호출 할 수 있는데 didSet으로 걸어두신 이유가 궁금해서요!
3. weak self는 어떤 기능을 하고있나요?
4. 저는 개인적으로 이렇게 코드가 있으면 stopAnimating의 짝궁이 어디있는지 궁금해져요ㅎㅎ

ttokin replied:
2. `products`에 데이터가 들어오자 마자 스냅샷을 찍고 로딩중 표시를 없애고 싶어서 didSet을 사용했어요. 여기가 아니라면 `fetchData()`의 `case .success`안에서 호출할 것 같아요. (`fetchData()` 안에서 호출하도록 수정해보았습니다.)
3. 여기서의 `weak self`는 `products`를 캡쳐하고 있습니다..!
4. 확실히 start-stop animating을 한눈에 보는게 가독성이 더 좋을 것 같습니다!

lina0322 commented on OpenMarket/OpenMarket/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,245 @@
+    
+    private let activityIndicator = UIActivityIndicatorView()
+    
+    private var dataSource: UICollectionViewDiffableDataSource<ProductListSection, Product>?
```
UICollectionViewDiffableDataSource는 어땠나요? 사용하기 편했나요?? 기존 방식과 어떤 차이가 있나요

ttokin replied:
제가 공부해보고 느낀 가장 큰 차이점은 아이템의 추가/삭제/이동이 일어났을 때 기존 데이터 소스의 경우 각 아이템의 index와 index path를 기반으로 셀의 내용을 업데이트 해야 하기 때문에 추가/삭제/이동시키기 위해서 `performBatchUpdates()` 메서드의 클로저를 직접 구현해야 하고, 이 과정에서 휴먼 에러로 인해 index 에러가 발생할 가능성이 높습니다. 하지만 Diffable Data Source의 경우 각 상태에 대한 스냅샷을 찍고, `applySnapshot()`을 호출만 해주면 스냅샷끼리 비교해서 변경사항을 업데이트 할 수 있습니다.

그런데 아직까지는 데이터를 불러와서 뷰를 그리기만 할 뿐, 데이터의 추가/삭제/이동을 구현하고 있지는 않아서 실제 구현 과정에서 큰 차이점을 느끼지는 못했던 것 같습니다..!

lina0322 commented on OpenMarket/OpenMarket/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,245 @@
+
+import UIKit
+
+final class OpenMarketViewController: UIViewController {
```
viewContoller가 너무 많은 일을 하고있어요~ 
(굳이 변경하진 않더라도) MVVM에 대해서 살짝 공부해봐도 좋을 것 같아요.

lina0322 commented on OpenMarket/OpenMarket/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,245 @@
+    }
+    
+    private func configureNavigationBar() {
+        navigationController?.navigationBar.scrollEdgeAppearance = navigationController?.navigationBar.standardAppearance
```
다른 Appearance는 설정하지 않아도 되나요?👀

ttokin replied:
6a62aa1fff2cb26ca07a3ba750fa141056344cbb

저걸 설정하게 된 이유는 스크롤 했을 때 내비게이션 바가 투명해지기 때문이었습니다. 찾아보니 iOS 15 이상에서는 업데이트 된 이후 내비게이션 바가 확장이 됨으로써 `scrollEdgeAppearance`는 기본적으로 투명한 배경으로 생성이 되고, 뒤에 컨텐츠가 없는 경우 기본적으로 투명한 배경색으로 보이게 변했다고 합니다 (관련 WWDC 영상도 시청 후 따로 내용 정리하겠습니다..!).
제 xcode 버전이 14.1 버전이어서 시뮬레이터들이 전부 iOS 16.1 버전으로 설정되어 있어, 빌드 세팅의 deployment target은 iOS 14지만 시뮬레이터는 iOS 16.1 버전으로 작동되어 내비게이션 바가 투명해지는 것이었습니다. 이외에 다른 문제점은 찾지 못해서 `scrollEdgeAppearance`만 설정했었습니다.
오늘 궁금해서 iOS 14 버전의 시뮬레이터를 다운받아서 실행해보니 아래 이미지와 같이 내비게이션뷰컨/리스트뷰의 배경색이 모두 검정색이고 `priceLabel`의 텍스트도 검정색이어서 `priceLabel`이 보이지 않는 문제가 새로 발생했습니다. 다크모드로 변경 시 `priceLabel`이 흰색 글씨로 변경되는 것을 보고 배경색 문제임을 알 수 있었습니다. 
![](https://i.imgur.com/jPlifp7.png)

배경색 문제를 해결하기 위해서는 `priceLabel`의 상위뷰, 만약 그 상위뷰가 투명하다면 상위뷰의 상위뷰에 배경색을 주어야 한다고 생각해서 리스트뷰/그리드뷰 마다 적절한 뷰에 `backgroundColor = .systemBackground`를 지정해 아래와 같이 다크모드 아닐때 / 다크모드 모두 잘 보이게 수정하였습니다.
- 리스트뷰
  | 다크모드 아닐 때 | 다크모드 |
  | -- | -- |
  | ![](https://i.imgur.com/z4yXcnF.png) | ![](https://i.imgur.com/YPtBKB7.png) |

- 그리드뷰
  | 다크모드 아닐 때 | 다크모드 |
  | -- | -- |
  | ![](https://i.imgur.com/WNbVqJm.png) | ![](https://i.imgur.com/EEGCLKn.png) |

lina0322 replied:
놀랍게도 흐름이 다크모드까지 갔네요. 오히려 좋아..👍
지금은 운이 좋게(?) 배경색만 변경했지만, label도 label color가 아니었다면 아마 수정했어야 하지 않을까 하는 추측을 해봅니다.

더불어
 > 배경색 문제를 해결하기 위해서는 priceLabel의 상위뷰, 만약 그 상위뷰가 투명하다면 상위뷰의 상위뷰에 배경색을 주어야 한다고 생각해서 리스트뷰/그리드뷰 마다 적절한 뷰에 backgroundColor = .systemBackground를 지정해 아래와 같이 다크모드 아닐때 / 다크모드 모두 잘 보이게 수정하였습니다.

요렇게 상위뷰와 하위뷰의 관계성에 대해서 파악한 부분도 좋았습니다👏

lina0322 commented on OpenMarket/OpenMarket/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,245 @@
+    
+    private func configureActivityIndicator() {
+        view.addSubview(activityIndicator)
+        activityIndicator.frame = CGRect(x: 0, y: 0, width: 100, height: 100)
```
이거 사이즈 키우면 커지나요~?👀

ttokin replied:
debug view hierarchy에서 확인해보니 아래와 같이 프레임의 크기만 커지고 컨텐트의 크기는 커지지 않네요..!
![](https://i.imgur.com/KxZ0fYi.png)

`activityIndicator.style = .large`를 활용해서 크기를 키워주는 것으로 변경했습니다!

lina0322 commented on OpenMarket/OpenMarket/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,245 @@
+        activityIndicator.startAnimating()
+        
+        let networkManager = NetworkManager()
+        networkManager.request(endpoint: OpenMarketAPI.productList(pageNumber: page, itemsPerPage: 20), dataType: ProductList.self) { result in
```
20이라는 숫자를 상수화하는게 좋겠어요. 
networkManager를 매번함수 호출때마다 생성하는게 나을까요? 아니면 들고있다가 사용하는게 나을까요..🧐

ttokin replied:
처음에 제가 생각했던 건 `networkManager`는 데이터를 불러올 때만 필요하니까 데이터를 불러오는 메서드 내에서 생성해서 사용하면 된다고 생각했었어요.
그런데 만약 `task`를 cancel 해야 할 상황이 생긴다면 같은 `networkManager`를 공유해야 하니까 메서드 밖에서 들고 있는 게 나을 것 같습니다.

lina0322 commented on OpenMarket/OpenMarket/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,245 @@
+                }
+                self.products += refinedProducts
+            case .failure(let error):
+                print(error)
```
고객들이 사용하는 앱에 print가 필요할까요?👀

ttokin replied:
ca2e302b43d7cd9975f82631bb7984e9395e9706

에러의 원인을 보여주는 alert를 보여주는 것으로 대체하였습니다!

lina0322 replied:
alert를 사용하는 것도 방법이고, debugPrint나 log 등을 사용할 수도 있겠어욤!

lina0322 commented on OpenMarket/OpenMarket/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,245 @@
+    }
+    
+    private func configureListCollectionView() {
+        let itemSize = NSCollectionLayoutSize(widthDimension: .fractionalWidth(1.0), heightDimension: .fractionalHeight(1.0))
+        let item = NSCollectionLayoutItem(layoutSize: itemSize)
+        item.contentInsets = NSDirectionalEdgeInsets(top: 5, leading: 5, bottom: 5, trailing: 5)
+        
+        let groupSize = NSCollectionLayoutSize(widthDimension: .fractionalWidth(1.0), heightDimension: .fractionalHeight(0.07))
+        let group = NSCollectionLayoutGroup.vertical(layoutSize: groupSize, subitems: [item])
```
되도록 상수화 해주세요~!

ttokin replied:
4deb0b900963ea0856f1917f98aa9d38efdbfe18

각각 설정할 수 있도록 열거형을 새로 생성해 주었습니다.

lina0322 commented on OpenMarket/OpenMarket/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,245 @@
+        
+        listCollectionView = UICollectionView(frame: .zero, collectionViewLayout: layout)
+        
+        guard let listCollectionView = listCollectionView else { return }
+        
+        view.addSubview(listCollectionView)
+        listCollectionView.delegate = self
+        
+        listCollectionView.translatesAutoresizingMaskIntoConstraints = false
+        NSLayoutConstraint.activate([
+            listCollectionView.topAnchor.constraint(equalTo: view.topAnchor),
+            listCollectionView.bottomAnchor.constraint(equalTo: view.bottomAnchor),
+            listCollectionView.leftAnchor.constraint(equalTo: view.leftAnchor),
+            listCollectionView.rightAnchor.constraint(equalTo: view.rightAnchor)
+        ])
```
위 내용과는 조금 다른 맥락이에요. 함수를 분리하는게 더 좋을거같네요!

ttokin replied:
4deb0b900963ea0856f1917f98aa9d38efdbfe18

gridView와 공통된 설정을 하는 부분을 메서드로 분리해 주었습니다!

lina0322 commented on OpenMarket/OpenMarket/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,245 @@
+        dataSource?.apply(snapshot, animatingDifferences: false)
+    }
+    
+    private func configureGridCollectionView() {
+        let layout = UICollectionViewFlowLayout()
+        layout.minimumLineSpacing = 10
+        layout.minimumInteritemSpacing = 5
+        layout.sectionInset = UIEdgeInsets(top: 10, left: 10, bottom: 10, right: 10)
+        gridCollectionView = UICollectionView(frame: .zero, collectionViewLayout: layout)
+        
+        guard let gridCollectionView = gridCollectionView else { return }
+        view.addSubview(gridCollectionView)
+        
+        gridCollectionView.delegate = self
+        
+        gridCollectionView.translatesAutoresizingMaskIntoConstraints = false
+        NSLayoutConstraint.activate([
+            gridCollectionView.topAnchor.constraint(equalTo: view.topAnchor),
+            gridCollectionView.bottomAnchor.constraint(equalTo: view.bottomAnchor),
+            gridCollectionView.leftAnchor.constraint(equalTo: view.leftAnchor),
+            gridCollectionView.rightAnchor.constraint(equalTo: view.rightAnchor)
+        ])
+    }
```
최대한 함수는 쪼개는게 좋아요~!

ttokin replied:
4deb0b900963ea0856f1917f98aa9d38efdbfe18

`listView`와 공통된 설정을 하는 부분을 메서드로 분리해 주었습니다!

lina0322 commented on OpenMarket/OpenMarket/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,245 @@
+    private func configureGridDataSource() {
+        guard let gridCollectionView = gridCollectionView else { return }
+        
+        gridCollectionView.register(GridCollectionViewCell.self, forCellWithReuseIdentifier: GridCollectionViewCell.identifier)
```
configureGridDataSource는 segmentControl이 변경될때마다 호촐될텐데 
매번 셀을 등록할 필요가 있나용??🫠

ttokin replied:
```swift
@objc private func segmentValueChanged(_ sender: UISegmentedControl) {
    guard let viewType: ViewType = ViewType(rawValue: sender.selectedSegmentIndex) else { return }

    switch viewType {
    case ViewType.list:
        configureListCollectionView()
        configureListDataSource()
        applySnapshot(for: products)
        listCollectionView?.isHidden = false
        gridCollectionView?.isHidden = true
    case ViewType.grid:
        configureGridCollectionView()
        configureGridDataSource()
        applySnapshot(for: products)
        gridCollectionView?.isHidden = false
        listCollectionView?.isHidden = true
    }
}
```

위 코드와도 관계가 있는 부분이라 생각되는데,
엘림의 말처럼 segment가 바뀔 때마다 뷰도 새로 그리고, 데이터소스도 새로 받아오고, 셀 설정도 전부 새로 하고 있는 것 같은데? 라는 생각이 들었어요. 실제로 그러고 있는게 맞을까요?

만약 그러고 있다면, segment가 바뀔 때 현재 표시하고 있는 뷰를 숨기고 / 새롭게 보여주어야 할 뷰가 이미 설정이 되어 있다면 그 뷰의 `isHidden`을 `false`로 바꾸어주는 동작만 해도 되지 않을까..? 라는 생각이 드는데 엘림의 생각은 어떤지 궁금합니닷..!

lina0322 replied:
ㅎㅎ네 저도 그렇게 생각합니다. 제가 의도한 방향이에요~👏

ttokin replied:
1537ec8b179cbbb4c7647fafe14bcd4a5e834e61 d8d303656447e5ab8d64e95f5f97c4de044a8882

로직을 조금 수정해보았습니닷..!!

lina0322 commented on OpenMarket/OpenMarket/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,245 @@
+        
+        dataSource = UICollectionViewDiffableDataSource<ProductListSection, Product>(collectionView: gridCollectionView) { (collectionView, indexPath, product) -> UICollectionViewCell? in
+            let cell = collectionView.dequeueReusableCell(withReuseIdentifier: GridCollectionViewCell.identifier, for: indexPath) as? GridCollectionViewCell
+            cell?.contentView.backgroundColor = .white
+            cell?.contentView.layer.borderColor = UIColor.gray.cgColor
+            cell?.contentView.layer.borderWidth = 1.0
+            cell?.contentView.layer.cornerRadius = 10.0
+            cell?.contentView.layer.masksToBounds = true
+            cell?.product = product
+            cell?.updateContents(product)
```
이정도면 guard처리를 하는게 낫겠어용ㅎㅎ
또한 contentView의 색상이나 이런것들을 변경하는건 cell 자체가 들고있는게 맞을 것 같아요
역할 분리가 필요해보입니답

ttokin replied:
오 안그래도 여기에 작성할지 셀에 작성할지 고민이 있었어요. 수정하겠습니다!

lina0322 commented on OpenMarket/OpenMarket/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,245 @@
+
+extension OpenMarketViewController: UICollectionViewDelegateFlowLayout {
+    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
+        let width: CGFloat = collectionView.frame.width / 2 - 15
+        let height: CGFloat = collectionView.frame.height / 3 - 20
```
의미를 알 수 없는 숫자들이 너무 많이 등장해요😭

ttokin replied:
4deb0b900963ea0856f1917f98aa9d38efdbfe18

코드가 많이 길어졌지만..! 상수를 이용해서 로직의 의도를 알 수 있도록 수정해 보았습니다.

lina0322 commented on OpenMarket/OpenMarket/SceneDelegate.swift:
```diff
@@ -7,45 +7,48 @@
-        // If using a storyboard, the `window` property will automatically be initialized and attached to the scene.
-        // This delegate does not imply the connecting scene or session are new (see `application:configurationForConnectingSceneSession` instead).
-        guard let _ = (scene as? UIWindowScene) else { return }
+        guard let scene = (scene as? UIWindowScene) else { return }
+        window = UIWindow(frame: scene.coordinateSpace.bounds)
+        window?.windowScene = scene
+        
+        let rootViewController = OpenMarketViewController()
+        window?.rootViewController = UINavigationController(rootViewController: rootViewController)
+        window?.makeKeyAndVisible()
```
scene은 무엇인가요? window의 역할은 어떤거죠?
AppDelegate와 SceneDelegate는 같은건가용?🫣

ttokin replied:
![](https://i.imgur.com/AqM7vMZ.png)
요론 느낌이라고 생각했습니다..!
동일한 앱을 여러 개 실행한다고 했을 때 앱의 각 인스턴스별 UI를 담당하는게 Scene이고, Scene위에 올라가는 것이 Window다..!
Window가 앱의 콘텐트를 담고 있는 역할을 하고, 실제로 콘텐트를 시각적으로 보여주는 것은 window의 뷰컨이 관리하는 뷰다..!
이렇게 이해했습니다...

lina0322 replied:
어머, 최고입니다👍
공부도 열심히 하고, 시각적 자료까지 만들다니 썸머캣의 노력ㅇㅔ 감동해버렸어요🥹

lina0322 commented on OpenMarket/OpenMarket/StockStatus.swift:
```diff
@@ -0,0 +1,10 @@
+//  OpenMarket
+//  Created by SummerCat on 2022/11/29.
+
+enum StockStatus: String {
+    case soldOut = "품절"
+    case enoughStock = "재고 여유 있음"
+    case remainingStock = "잔여수량"
+    case stockError = "오류"
+}
```
LGTM👍

lina0322 commented on OpenMarket/OpenMarket/View/GridCollectionViewCell.swift:
```diff
@@ -0,0 +1,175 @@
+import UIKit
+
+final class GridCollectionViewCell: UICollectionViewCell {
+    static let identifier = "gridCell"
```
identifier를 이렇게 수동으로 설정하지 않고, 자동으로 설정할 순 없을까요?👀
(힌트: String, self)

ttokin replied:
`static let identifier: String = String(describing: GridCollectionViewCell.self)`와 같이 수정해 보았습니다!

lina0322 replied:
`static let identifier: String = String(describing: GridCollectionViewCell.self)`
여기서 GridCollectionViewCell을 제외 시킬 수 있을까요?
또한 cell이 여러개일때 복붙하기보다는, 하나의 공통 코드로 만들어서 편하게 적용시킬 수 있는 방법은 뭐가 있을까용?!

ttokin replied:
`GridCollectionViewCell`을 빼고 `self`만 써도 작동은 하는데 노란색 오류창이 떠서 신경쓰여서요.. 뭔가 다른 방법이 있을까요? String의 이니셜라이저들을 열심히 읽어봤는데.. 잘 모르겠어요😭

lina0322 replied:
노란색 경고창의 내용이 무엇이었나요ㅎㅎ

ttokin replied:
![](https://i.imgur.com/iDrM6OV.png)
요고입니다 엘림..!

ttokin replied:
identifier를 연산 프로퍼티로 변경해주니 위 경고창이 뜨지 않네요🤔
일반 프로퍼티의 경우 인스턴스가 fully initialized 되기 전일 수 있어서 경고창이 뜨는 것 같고, 연산 프로퍼티의 경우 해당 프로퍼티에 접근할 때 `get { }` 안의 코드가 실행되어 `{ }`안에서 값을 반환해주기 때문에 경고창이 발생하지 않는 것 같습니다.

lina0322 replied:
ㅎㅎㅎ좋습니다. 역시 방법은 있으니까요! 아주 잘 찾았습니다.
이제는 이 부분을 protocol로 만들어보는거에요~~ 화이팅

ttokin replied:
c1120cfa34feae12c934021b6e3dba95d45de1ff

수정 완료입니다! 👍🏻

lina0322 commented on OpenMarket/OpenMarket/View/GridCollectionViewCell.swift:
```diff
@@ -0,0 +1,175 @@
+
+import UIKit
+
+final class GridCollectionViewCell: UICollectionViewCell {
+    static let identifier = "gridCell"
+    var product: Product?
+    
+    required init?(coder: NSCoder) {
+        super.init(coder: coder)
+    }
+    
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        
+        configureUI()
+    }
```
두 init의 차이는 무엇인가요? 왜 coder 쪽에는 내용이 없나요..?!

ttokin replied:
`init?(coder: NSCoder)`는 스토리보드에서 받아온 정보를 사용해서 뷰를 initialize하는 메서드이고, `init(frame: CGRect)`는 코드로 뷰를 직접 그려서 initialize하는 메서드로 알고 있습니다.
`init?(coder: NSCoder)`에는 `fatalError`를 찍어줄 수도 있고 제가 구현한 것과 같이 `super.init()`을 해줄 수도 있다고 알고 있습니다. 둘의 차이는 `fatalError`를 타게 되면 앱이 죽고, `super.init()`을 타면 앱이 죽지는 않으나 화면에 아무것도 보이지 않게 되는 것으로 알고 있습니다.
뷰를 코드로 그리고 있기 때문에 `init?(coder: NSCoder)`가 필요하지 않지만, `required` 이니셜라이저이기 때문에 반드시 구현해주어야 합니다.

lina0322 replied:
> super.init()을 타면 앱이 죽지는 않으나 화면에 아무것도 보이지 않게 되는 것으로 알고 있습니다.

왜 화면에 아무것도 안보일까요?? configureUI()를 마찬가지로 해주면 보이지 않을까용?👀

ttokin replied:
`init?(coder: NSCoder)`가 실패 가능한 이니셜라이저여서 셀의 인스턴스를 생성시키지 못하기 때문에(매개변수로 받아야 할 NSCoder 타입의 매개변수를 받지 못했기 때문) `init?(coder:NSCoder)` 내에서 `configureUI()`를 호출해도 뷰가 그려지지 않는 것 같습니다!
(`init(frame: CGRect)`를 주석처리하고 실험해보니 화면이 아예 안 뜨더라구요)

lina0322 replied:
누군가 스토리보드로 이 셀을 호출해도 마찬가지일까요??
만약에 스토리보드로 구현을 막을거라면 fatal로 처리하는건 어떨까요?
혹은 누군가 스토리보드로 구현을 할 수 있도록 똑같이 처리를 해주는건 어떨까요?

ttokin replied:
c1120cfa34feae12c934021b6e3dba95d45de1ff

아항 지금 상태로 두면 모르는 사람이 보았을 때 확실히 헷갈릴 수 있겠네요..!
스토리보드에서 구현하지 못하게 하도록 `fatalError`로 변경해주었습니다.

lina0322 commented on OpenMarket/OpenMarket/View/GridCollectionViewCell.swift:
```diff
@@ -0,0 +1,175 @@
+    }
+    
+    override func prepareForReuse() {
+        super.prepareForReuse()
```
prepareForReuse가 되는 타이밍은 언제일까요??

ttokin replied:
새로운 셀을 만들기 위해 `dequeueReusableCell`을 호출하면, 재사용가능한 셀이 있을 경우 `dequeueReusableCell` 메서드가 호출합니다.
![](https://i.imgur.com/hBhV6ZD.png)

lina0322 replied:
LGTM😻

lina0322 commented on OpenMarket/OpenMarket/View/GridCollectionViewCell.swift:
```diff
@@ -0,0 +1,175 @@
+        label.textAlignment = .center
+        return label
+    }()
+    
+    private let priceLabel: UILabel = {
+        let label = UILabel()
+        label.font = .preferredFont(forTextStyle: .body)
+        label.textAlignment = .center
+        label.textColor = .gray
+        return label
+    }()
+    
+    private let bargainPriceLabel: UILabel = {
+        let label = UILabel()
+        label.font = .preferredFont(forTextStyle: .body)
+        label.textAlignment = .center
+        label.textColor = .gray
+        return label
+    }()
+    
```
동일한 라벨인가요??👀
비슷한 라벨을 생성한다면 함수를 통해서 생성하는건 어떨까요

ttokin replied:
97738f1572ce40741f71f6245d2da84b85f8ed7e

`UILabel`의 extension으로 label을 생성하는 함수를 구현해 주었습니다.
처음에는 cell에서 채택하는 프로토콜의 함수로 구현했다가, 이렇게 구현할 경우 모든 label을 lazy var로 생성해야 해서 UILabel의 extension에 구현하는 것으로 변경해 주었습니다.

lina0322 commented on OpenMarket/OpenMarket/View/GridCollectionViewCell.swift:
```diff
@@ -0,0 +1,175 @@
+    }
+    
+    func updateContents(_ product: Product) {
+        DispatchQueue.global().async {
+            guard let url = URL(string: product.thumbnailURL),
+                  let data = try? Data(contentsOf: url),
+                  let image = UIImage(data: data) else { return }
+            DispatchQueue.main.async {
+                if product == self.product {
+                    self.productImageView.image = image
+                    self.productNameLabel.text = product.name
+                    self.updatePriceLabel(product)
+                    self.updateStockLabel(product)
+                }
+            }
+        }
```
왜 Data(contentsOf: url)을 사용했나요?!

ttokin replied:
b1819899412aa541cda6647c46b2d8c658e250c7

UIImage를 만드는 방법을 먼저 검색해보고, Data 타입의 매개변수를 받아서 init해야 한다는 것을 알아내고, 그 다음에 url을 사용해서 data를 만드는 방법을 검색하는 흐름으로 갔던 것 같습니다.
그런데 해당 메서드의 [공식문서](https://developer.apple.com/documentation/foundation/nsdata/1413892-init)를 보니 네트워킹하는 url일 경우 절대 이 메서드를 사용하지 말라고 되어있네요..! (네트워크에 문제가 있을 경우 앱이 죽을 수 있기 때문에)
공식문서에서 권장하는 대로 URLSession의 `dataTask` 함수를 사용하는 방향으로 수정했습니다!

lina0322 commented on OpenMarket/OpenMarket/View/GridCollectionViewCell.swift:
```diff
@@ -0,0 +1,175 @@
+            return
+        }
+        
+        if product.stock > 10000 {
```
의미를 알 수 없는 숫자들이 참 많아요~~
참고로 이런건 10_000 이렇게 쓰는게 더 좋아요~!

ttokin replied:
ca2e302b43d7cd9975f82631bb7984e9395e9706

`Int`의 extension을 사용해 제거해 주었습니다!

lina0322 commented on OpenMarket/OpenMarket/View/ListCollectionViewCell.swift:
```diff
@@ -0,0 +1,203 @@
+
+import UIKit
+
+final class ListCollectionViewCell: UICollectionViewCell {
```
아까 셀이랑 매우 유사하네요.
같은 부분을 하나로 합쳐서 상속받거나 프로토콜을 채택할 수도 있지 않을까요??

ttokin replied:
b1819899412aa541cda6647c46b2d8c658e250c7

두 셀에서 공통으로 가진 요소와 함수를 분리해내서 프로토콜화 시켜주었습니다.

lina0322 reviewed (changes requested):
썸머캣 반가워요ㅎㅎㅎㅎㅎ
얼굴을 봐서 그런가 더 친근하네요! 일단 코멘트 먼저 달아드렸어요
이 부분 해결하면서, 앞서서 고민했던것도 함께 생각하고 고민해봐요~
해결이 잘 안 되거나 어려운거 이해 안되는건 편하게 DM 주시고요!
남겨주신 조언이 필요하거나 궁금한것들은 지금 남긴 코멘트들 하다보면 어느정도 윤곽이 잡힐 것 같아서 우선 패스할게요~ㅎㅎ

화이팅!!!

(개인적으로 UI에 대해 고민한 부분 아주 좋았어요😉)

ttokin committed:
feat: SwiftLint 적용

ttokin committed:
chore: Deployment Target 14.0으로 통일, style: 코드 컨벤션과 가독성을 고려한 개행 수정

ttokin committed:
refactor: 뷰와 셀에 배경색 설정

ttokin committed:
refactor: Diffable Data Source와 Snapshot에서 Product의 인스턴스가 아닌 Product.ID를 저장하도록 수정

ttokin committed:
refactor: segmentedControl 설정 메서드로 추가, segmentedControl의 item 추가 메서드 생성

ttokin committed:
refactor: cell의 identifier 정의 로직 수정, activityIndicator 크기 지정

ttokin committed:
refactor: networkManager의 생성 위치 변경, dataTask 함수에 @Sendable 추가

ttokin committed:
refactor: CollectionViewCellType 프로토콜 생성, ListCell과 GridCell에서 CollectionViewCellType 채택, 이미지를 url에서 가져오는 부분 dataTask 메서드를 통해 가져오도록 수정

ttokin committed:
refactor: UILabel extension 구현, createLabel 메서드 구현

ttokin committed:
refactor: LayoutConstants 열거형 생성, listView와 gridView 공통 설정 부분 메서드로 분리

ttokin committed:
refactor: applySnapshot, stopAnimating 호출 위치 변경

ttokin committed:
refactor: AlertConstant, Int extension 구현, StockStatus 열거형 rawValue 추가, 매직 리터럴 제거

lina0322 commented on OpenMarket/.swiftlint.yml:
```diff
@@ -0,0 +1,7 @@
+disabled_rules:
+  - trailing_whitespace
+  - empty_enum_arguments
+  - xctfail_message
+  - line_length
+  - unused_closure_parameter
+  - legacy_constructor
```
린트 넘 편하죠?ㅎㅎ 어떤게 젤 좋은 룰이었나요

ttokin replied:
가장 많이 썼던 건 세로로 엔터 여러번  친 것 잡아주는 거랑 TODO를 워닝으로 보여주니 못보고 지나칠 수 있는 TODO를 까먹지 않고 확인할 수 있어서 좋았습니다!

lina0322 commented on OpenMarket/OpenMarket.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist:
```diff
@@ -0,0 +1,8 @@
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+	<key>IDEDidComputeMac32BitWarning</key>
```
오홍 이건 왜 설정했나요??

ttokin replied:
제가 수동으로 Info.plist 파일에 무언가를 설정해 준 기억은 없는데욥.... 아마 SwiftLint 적용하다가 Xcode에서 뜨는 노란경고창 추천 설정을 눌렀다가 자동으로 설정된 것이 아닐까....추측해봅니다...ㅠㅠ
찾아보니 Info.plist 설정과 관계가 있는 파일은 맞는 것 같습니다🥲

lina0322 commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,274 @@
+        configureListCollectionView()
+        configureListDataSource()
+        
+        configureActivityIndicator()
```
전체적으로 줄 바꿈의 기준을 제가 짐작하기 어렵네요😅
줄 바꿈도 의도가 있어야 한다는것 잊지 않고있죠?

ttokin replied:
d8d303656447e5ab8d64e95f5f97c4de044a8882

서로 관련이 크다고 생각한 것들은 묶어주고, 메서드 형태로 정의된 것은 줄바꿈해주고 이런 식으로 작성하긴 했는데 제가 봐도 난잡해서 어떻게 정리해야 할지... 고민이 많았습니다..... (함수도 마찬가지로요..!)
프로퍼티들은 뷰인 것 / 뷰가 아닌 것으로 나누어 보았고, 함수도 묶어보았습니다.

lina0322 commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,274 @@
+        configureListCollectionView()
+        configureListDataSource()
+        
+        configureActivityIndicator()
```
configureUI, setupUI 등으로 관련있는 함수들을 한 번 더 묶으면 어떨까요??

ttokin replied:
d8d303656447e5ab8d64e95f5f97c4de044a8882

뷰와 관련된 함수들을 묶어보았습니다..!

lina0322 commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,274 @@
+        navigationItem.titleView = segmentedControl
+    }
+
+    private func addSegment(with title: String?, at index: Int) {
+        segmentedControl?.insertSegment(withTitle: title, at: index, animated: false)
+    }
```
어떻게 보면 segmentedControl의 함수를 거의 그대로 사용한 것 처럼 보이기도 하는데요ㅎㅎ
이렇게 감싼건 어떤 장점때문인가요??

ttokin replied:
ba0a497377c6a211e14dae7aa25b7aa1b37d5586

지금 당장은 segment를 추가해줄 일이 없지만, 엘림이 지난번에 segment를 추가하고 싶은 일이 생긴다면 세그먼트컨트롤의 아이템을 하드코딩 해야 한다는 점을 지적해주셨어서 메서드로 구현해두면 쉽게 처리할 수 있지 않을까..? 생각했는데
어차피 새로 구현한 메서드를 호출하는 것이든, 기본 구현된 함수를 호출하는 것이든 작업은 똑같겠네요..!
삭제 처리했습니다.

lina0322 commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,274 @@
+            applySnapshot(for: products)
+            gridCollectionView?.isHidden = false
+            listCollectionView?.isHidden = true
+        }
```
지난 코멘트에서 얘기 나눴던 부분 적용 해봅시다~

ttokin replied:
1537ec8b179cbbb4c7647fafe14bcd4a5e834e61 d8d303656447e5ab8d64e95f5f97c4de044a8882

수정 해보았습니다!

lina0322 commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,274 @@
+            productRegisterView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
+            productRegisterView.trailingAnchor.constraint(equalTo: view.trailingAnchor)
+        ])
+    }
```
`+ 버튼을 통해 새로운 상품을 등록하는 화면으로 이동할 수 있습니다.`에서 
빈 화면을 addSubview의 느낌으로 이해한건가요?👀

ttokin replied:
559c9eac1de30a4a395335b8c99af7668fcf9a14

잠깐 사탄 들렸었나봐요... 새로운 뷰 컨트롤러를 모달로 띄우도록 수정 완료입니다.

lina0322 commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,274 @@
+            switch result {
+            case .success(let productList):
+                var refinedProducts: [Product] = []
+                for product in productList.products {
+                    if self.products.contains(product) { continue }
+                    refinedProducts.append(product)
+                }
```
for in where이나 map 등을 활용할 수 있을까요??

ttokin replied:
b36848c7da6404c60af65536ba8e407ecb3551f4

for in where은 이번에 엘림이 이야기해주셔서 처음 알게되었어요!
활용해서 수정해보았습니다.

lina0322 commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,274 @@
+                self.applySnapshot(for: self.products)
+                self.activityIndicator.stopAnimating()
+            case .failure(let error):
+                self.showDataRequestFailureAlert(error)
```
실패하면 activityIndicator는 안멈추나요??

ttokin replied:
b36848c7da6404c60af65536ba8e407ecb3551f4

멈추도록 하는 게 맞는 거 같습니다..! 수정 완료입니다.

lina0322 commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,274 @@
+        networkManager.request(endpoint: OpenMarketAPI.productList(pageNumber: page, itemsPerPage: 20), dataType: ProductList.self) { result in
+            switch result {
+            case .success(let productList):
+                var refinedProducts: [Product] = []
+                for product in productList.products {
+                    if self.products.contains(product) { continue }
+                    refinedProducts.append(product)
+                }
+                self.products += refinedProducts
+                self.applySnapshot(for: self.products)
```
내부 함수를 분리하는게 더 좋을 것 같아요~

ttokin replied:
b36848c7da6404c60af65536ba8e407ecb3551f4

수정 완료입니다!

lina0322 commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,274 @@
+            cell.updateImage(product)
+        }
+        
+        dataSource = UICollectionViewDiffableDataSource<ProductListSection, Product.ID>(collectionView: listCollectionView) { colllectionView, indexPath, identifier -> UICollectionViewCell? in
```
이 줄 너무 길어요~ 줄바꿈 해봅시당

ttokin replied:
b36848c7da6404c60af65536ba8e407ecb3551f4

수정 완료입니다!

lina0322 commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,274 @@
+        
+        dataSource = UICollectionViewDiffableDataSource<ProductListSection, Product.ID>(collectionView: listCollectionView) { colllectionView, indexPath, identifier -> UICollectionViewCell? in
+            
+            var product: Product?
+            self.products.forEach {
+                if $0.id == identifier { product = $0 }
+            }
```
id값을 넘겨서 product를 반환하는 함수를 하나 만들어서 써봅시당

ttokin replied:
b36848c7da6404c60af65536ba8e407ecb3551f4

두 곳에서 같은 로직이 반복되고 있으니 함수화 하는게 좋겠네요! 반영했습니다!

lina0322 commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,274 @@
+    }
+    
+    private func applySnapshot(for items: [Product]?) {
+        var itemIdentifiers: [Product.ID] = []
+        products.forEach {
+            itemIdentifiers.append($0.id)
+        }
```
Product.ID를 사용하는게 전반적으로 불필요한 코드가 늘어나는 기분이 계속 들긴하네요ㅎㅎ;;😅

ttokin replied:
맞아요... 공식문서에서는 ID를 사용하는게 peak performance를 고려하면 더 좋다고 하는데.. 조금 의문이네요....
WWDC라도 다시 보면서 해당 내용을 확인해보ㅏ야 할 것 같아요.

lina0322 commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,274 @@
+            
+            let cell = collectionView.dequeueReusableCell(withReuseIdentifier: GridCollectionViewCell.identifier, for: indexPath) as? GridCollectionViewCell
+            cell?.product = product
+            cell?.updateContents(product)
+            cell?.updateImage(product)
```
cell이 이미 product을 들고있는거 아닌가용? 왜 또 넘겨줘야하나요~?
아니면 넘겨준 함수 내부에서 product를 cell 안에 담아줄 수도 있을 것 같은데요!

ttokin replied:
76f1ac8299cfc76b3df61a02adddbde400356ead

cellRegistration에서 cell을 세팅할 수 있다고 해서 여기서 또 넘겨줬던 것 같아요..!
아래 dequeue하는 부분에서 product를 할당하는 부분을 지우고, update하는 함수들은 프로퍼티 옵저버를 활용해 처리하도록 변경했습니다.

lina0322 commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,274 @@
+
+extension OpenMarketViewController: UICollectionViewDelegateFlowLayout {
+    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
+        let width: CGFloat = collectionView.frame.width / (LayoutConstants.gridPerRow.value - LayoutConstants.gridCellMinimumInteritemSpacing.value * (LayoutConstants.gridPerRow.value + 1))
+        let height: CGFloat = collectionView.frame.height / (LayoutConstants.gridPerCol.value - LayoutConstants.gridCellMinimumInteritemSpacing.value * (LayoutConstants.gridPerCol.value + 1))
```
줄바꿈하는 습관이 필요해요~
<img width="1490" alt="스크린샷 2022-12-07 22 09 25" src="https://user-images.githubusercontent.com/49546979/206187626-71599a52-2a79-4805-ad60-148bf26db2f0.png">
한눈에 다 안들ㅇㅓ와요ㅠ.ㅠ

ttokin replied:
b36848c7da6404c60af65536ba8e407ecb3551f4

수정 완료입니다ㅠㅠ! 이렇게 바꿔도 되는지 잘 모르겠어요...

lina0322 commented on OpenMarket/OpenMarket/Network/NetworkError.swift:
```diff
@@ -4,9 +4,9 @@
-    case noData
-    case URLError
-    case decodingError
+enum NetworkError: String, Error {
+    case statusCodeError = "http 응답 에러"
+    case noData = "데이터 없음"
+    case URLError = "URL 오류"
+    case decodingError = "디코딩 오류"
 }
```
`networkError.rawValue` 말고 `networkError.errorMessage` 이런식으로 접근할 수 있는 방법은 어떨까요?

ttokin replied:
aa617c8199e9205f2b285f0ff71a79878db738f1

rawValue보다 description이 더 직관적이라고 생각해서 description 프로퍼티를 새로 구현해서 처리했습니다!

lina0322 commented on OpenMarket/OpenMarket/Network/URLSessionable.swift:
```diff
@@ -5,5 +5,5 @@
 
 protocol URLSessionable {
-    func dataTask(with request: URLRequest, completionHandler: @escaping ((Data?, URLResponse?, Error?) -> Void)) -> URLSessionDataTask
+    func dataTask(with request: URLRequest, completionHandler: @escaping @Sendable (Data?, URLResponse?, Error?) -> Void) -> URLSessionDataTask
```
`@Sendable`은 어떤 키워드인가욤👀

ttokin replied:
Concurrent하게 작업하기 위해 (다른 스레드에서 작업하기 위해) 값을 넘겨줄 때, Sendable 프로토콜을 준수하거나 `@Sendable` attribute이 붙은 경우 캡쳐하는 값들을 값 타입으로 다른 스레드로 보내기 때문에, thread-safety를 보장할 수 있습니다.

추가했던 이유는 `@Sendable` 키워드를 포함하는 형태로 정의된 함수가 이미 존재하고 있기 때문에 발생하는 오류를 정정하기 위해 추가했었습니다.

lina0322 commented on OpenMarket/OpenMarket/View/CollectionViewCellType.swift:
```diff
@@ -0,0 +1,84 @@
+
+import UIKit
+
+protocol CollectionViewCellType: AnyObject {
```
AnyObject는 어떤 키워드인가요??

ttokin replied:
해당 프로토콜을 class에서만 상속할 수 있도록 하기 위한 키워드입니다!

lina0322 commented on OpenMarket/OpenMarket/View/CollectionViewCellType.swift:
```diff
@@ -0,0 +1,84 @@
+import UIKit
+
+protocol CollectionViewCellType: AnyObject {
+    static var identifier: String { get }
```
오오 여기 잘 적용했네요! 보통 이렇게 잘 사용하는 편이라, 더 많은 셀에도 사용될 수 있게 분리하는 것도 좋겠어요👏

lina0322 commented on OpenMarket/OpenMarket/View/CollectionViewCellType.swift:
```diff
@@ -0,0 +1,84 @@
+    var bargainPriceLabel: UILabel { get }
+}
+
+extension CollectionViewCellType {
```
protocol로 묶어서 사용해보니 어떠세용?👀

ttokin replied:
중복되는 부분이 싹 빠져서 각 cell 파일의 가독성이 많이 좋아진 것 같습니다 =) 너무 행복해요

lina0322 commented on OpenMarket/OpenMarket/View/CollectionViewCellType.swift:
```diff
@@ -0,0 +1,84 @@
+    
+    private func fetchImage(_ url: URL, completion: @escaping (UIImage) -> Void) {
+        task = URLSession.shared.dataTask(with: url) { data, response, error in
+            guard error == nil else { return }
+            
+            guard let response = response as? HTTPURLResponse,
+                  (200..<300).contains(response.statusCode) else { return }
+            
+            guard let data else { return }
+            
+            guard let image = UIImage(data: data) else { return }
```
왜 guard문을 각각 나눠서 작성했나요?

ttokin replied:
ba0a497377c6a211e14dae7aa25b7aa1b37d5586

제가 잠시 미쳤었나봐요......... 수정 완료입니다😥
앞으로는 더 꼼꼼히 검토한 후 커밋하겠습니다.

lina0322 commented on OpenMarket/OpenMarket/View/Constants/LayoutConstants.swift:
```diff
@@ -0,0 +1,29 @@
+//  OpenMarket
+//  Created by SummerCat on 2022/12/06.
+
+enum LayoutConstants {
+    case listCellContentInset
+    case gridCellSectionInset
+    case gridCellMinimumLineSpacing
+    case gridCellMinimumInteritemSpacing
+    case gridPerRow
+    case gridPerCol
+    
+    var value: Double {
+        switch self {
+        case .listCellContentInset:
+            return 5
+        case .gridCellSectionInset:
+            return 10
+        case .gridCellMinimumLineSpacing:
+            return 10
+        case .gridCellMinimumInteritemSpacing:
+            return 5
+        case .gridPerRow:
+            return 2
+        case .gridPerCol:
+            return 3
+        }
+    }
+}
```
음, 이걸 앱 전체가 알 필요가 있을까요? 사용하는 곳에서 각각의 상수를 들고 있는게 더 좋을 것 같아서요ㅎㅎ
이런 상수들을 enum으로 갖는것과 struct로 갖는건 어떤 차이가 있을까요?

ttokin replied:
29acdc85b7f14d2e7528314808310366cbc2afba

다시 확인해보니 OpenMarketVC에서만 알고 있으면 되겠네요..! 수정 완료입니다.

struct로 네임스페이스를 만들 경우에는 해당 struct가 필요하든 필요하지 않든 다른곳에서 어쨌든 해당 struct의 인스턴스를 만들 수 있도록 접근 가능해지고, 만약 인스턴스를 만들지 못하도록 방지하려면 구조체 안에 `private init() { }`을 추가해줘야 하는데,
enum으로 네임스페이스를 만들면 애초에 인스턴스화가 불가능하기 때문에 굳이 private initializer를 만들어 줄 필요도 없어서 더 간결해직고, 실수로 인스턴스화 될 일도 없기 때문이라고 생각합니다.

lina0322 commented on OpenMarket/OpenMarket/View/GridCollectionViewCell.swift:
```diff
@@ -0,0 +1,114 @@
+            stockLabel.heightAnchor.constraint(equalTo: cellStackView.heightAnchor, multiplier: 0.1)
+        ])
+        
+        contentView.backgroundColor = .systemBackground
+        contentView.layer.borderColor = UIColor.gray.cgColor
+        contentView.layer.borderWidth = 1.0
+        contentView.layer.cornerRadius = 10.0
+        contentView.layer.masksToBounds = true
+    }
```
함수 분리하는 습관이 필요해요~!

ttokin replied:
29acdc85b7f14d2e7528314808310366cbc2afba

컨텐트뷰(그리드)를 그리는 함수만 분리해주었습니다.

lina0322 reviewed (commented):
썸머캣~~ 정말 고생이 많았어용!!👏
더 많은 코멘트를 달아버렸나 싶긴하지만 일단 남겨봅니당ㅎㅎㅎㅎ
여기서 계속 수정해도 좋고, 머지를 원하면 머지하고 다음거랑 함께 진행해도 좋아요~
편하게 말씀해주세용!
이번ㅇㅔ도 화이팅>_<

ttokin committed:
refactor: collectionView와 dataSource를 설정하는 함수 호출하는 로직 수정, gridCell 크기 계산 로직 오류 수정

ttokin committed:
refactor: cell의 identifier 프로퍼티 프로토콜에서 정의하도록 수정, 스토리보드로 셀 생성 시 fatalError 처리하도록 수정

ttokin committed:
refactor: dataSource를 listDataSource, gridDataSource로 분리, applySnapshot 메서드 수정, configureUI 메서드 생성

ttokin committed:
refactor: addSegment 함수 제거, 오토레이아웃 제약 수정

ttokin committed:
refactor: productRegisterView 제거 후 ProductRegisterVC 구현

ttokin committed:
refactor: filterProducts, findProduct 메서드 생성

ttokin committed:
refactor: updateImage, updateContents 함수 호출 위치 변경

ttokin committed:
refactor: LayoutConstraints 정의 위치 변경, configureGrid 메서드 분리

ttokin committed:
refactor: NetworkError에 대한 설명을 description 프로퍼티로 제공하는 것으로 변경

ttokin committed:
refactor: 네트워킹 API 수정, request 메서드 역할 수정

- HTTPMethod 열거형 생성
- Endpointable, OpenMarketAPI  제거
- request 메서드에서 dataTask
- request 메서드에서 데이터 디코딩 분리, URLSessionable 프로토콜 메서드로 변경

ttokin committed:
feat: fetchProductDetail 메서드 구현

ttokin committed:
refactor: createURLRequest 메서드의 httpMethod 매개변수 제거

lina0322 commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,290 @@
+        configureUI()
+        configureListDataSource()
+        
+		fetchProductList(page: pageNumber)
```
앞에 간격 무슨일이에여!
쭉 보니까 전체적으로 린트 설정이나 Xcode 설정 잘못된거같아요ㅠㅠ

dev-summer replied:
로칼에선 문제없이 보였는데 깃헙 업로드 버전을 보니 인덴트가 이상하네요...🫠 중간에 Xcode 인덴트 설정을 space -> tab으로 한번 바꿨다가 저기만 저렇게 된 것 같아요.. 수정 완료입니다!

lina0322 commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,290 @@
+        switch viewType {
+        case ViewType.list:
+            if listCollectionView == nil {
+                configureListCollectionView()
+                configureListDataSource()
```
configureUI()에서 하고 있는데 또 해야하나용?

dev-summer replied:
list-grid간에 전환하면서 `listCollectionView`나 `listDataSource`가 deinit될 일이 없어 보이기는 하지만 혹시 모른다고 생각해서 작성했었어요. 요것도 제거하였습니다 =)

lina0322 commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,290 @@
+                configureGridCollectionView()
+                configureGridDataSource()
+            }
+            applySnapshot(for: products)
```
applySnapshot도 왔다갔다 할때마다 하는게 맞을까요??

dev-summer replied:
`applySnapshot`에서 두 데이터소스에 모두 apply하고 있으니 왔다갔다 할 때마다 할 필요가 없어졌네욤..!
제거 완료입니다.

lina0322 commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,290 @@
+        activityIndicator.center = view.center
+    }
+
+	private func fetchProductList(page: Int) {
+        activityIndicator.startAnimating()
+		networkManager.fetchProductList(page) { result in
+			switch result {
+			case .success(let productList):
+				self.filterProducts(productList.products)
+				DispatchQueue.main.async {
+					self.applySnapshot(for: self.products)
+					self.activityIndicator.stopAnimating()
+				}
+			case .failure(let error):
+				self.showDataRequestFailureAlert(error)
+				DispatchQueue.main.async {
+					self.activityIndicator.stopAnimating()
+				}
+			}
+		}
+    }
```
이게 무슨일이람~~ PR 보내고 나서도 changes에서 확인 하는 습관이 필요해요!

lina0322 commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,290 @@
+extension OpenMarketViewController: UICollectionViewDelegate {
+    func collectionView(_ collectionView: UICollectionView, willDisplay cell: UICollectionViewCell, forItemAt indexPath: IndexPath) {
+        if indexPath.item == products.count - 1 {
+            pageNumber += 1
+			fetchProductList(page: pageNumber)
```
린트 제대로 동작하고 잇나여!!

lina0322 commented on OpenMarket/OpenMarket/Controller/ProductRegisterViewController.swift:
```diff
@@ -0,0 +1,15 @@
+
+import UIKit
+
+class ProductRegisterViewController: UIViewController {
```
final 키워드 확인하고 붙여주십숑

lina0322 commented on OpenMarket/OpenMarket/Network/NetworkManager.swift:
```diff
@@ -68,4 +29,48 @@ final class NetworkManager {
+		}
+	}
+	
+	// TODO: - POST
+	func registerProduct() {
+		let endpoint = Endpoint.registerProduct
+		guard var urlRequest = endpoint.createURLRequest() else { return }
+		urlRequest.httpMethod = HTTPMethod.post.rawValue
+
+		
+	}
+	
+	// TODO: - PATCH
+	
+	// TODO: - DELETE
```
PATCH랑 DELETE도 나중에 해봅시당~!

dev-summer replied:
넵!! POST PATCH DELETE 모두 방학기간동안 뿌셔보겠습니닷💪

lina0322 reviewed (changes requested):
썸머캣! 전체적으로 지금 앞쪽 간격 설정을 잘못한거같아요ㅠ 한번확인해주세요~
긴 기간동안 고생 많았고, 진짜진짜 많이 발전해서 보는 제가 뿌듯하네요
마지막까지 리팩토링 하느라고 정말정말 고생했습니다! 다음에는 MVVM 공부해서 또 리팩해봅시다
수고하셨어요~!! 🙇‍♀️

dev-summer committed:
Update README.md

STEP2 타임라인까지 업데이트

dev-summer committed:
Update README.md

UML, 실행 화면 수정

dev-summer committed:
refactor: 불필요하게 호출되는 메서드 제거

dev-summer committed:
chore: indent 수정

dev-summer committed:
chore: 전체 indent 수정
